mirgecom/logging_quantities.py

```diff
@@ -0,0 +1,247 @@
+            if candidate.is_file() and "emirge" in candidate.open().read():
+                path_to_version_sh = str(candidate)
+                break
+
+    if path_to_version_sh is None:
+        from warnings import warn
+        warn("Could not find emirge's version.sh. No package versions recorded.")
+        mgr.set_constant("emirge_package_versions",
+            "Could not find emirge's version.sh. No package versions recorded.")
+
+    output = os.popen(path_to_version_sh).read()
+    if output == "":
+        output = "Could not record emirge's package versions."
+
+    mgr.set_constant("emirge_package_versions", output)
+
+
+def set_state(mgr: LogManager, state: ndarray) -> None:
+    """Update the state of all :class:`StateConsumer` of the log manager `mgr`."""
+    for gd_lst in [mgr.before_gather_descriptors,
+            mgr.after_gather_descriptors]:
+        for gd in gd_lst:
+            if isinstance(gd.quantity, StateConsumer):
+                gd.quantity.set_state(state)
+
+
+class StateConsumer:
+    """Base class for quantities that require a state for logging."""
+
+    def __init__(self):
+        self.state = None
+
+    def set_state(self, state: ndarray) -> None:
+        """Set the state of the object."""
+        self.state = state
+
+
+class PhysicalQuantity(LogQuantity, StateConsumer):
+    """Logging support for physical quantities."""
+
+    def __init__(self, discr: Discretization, quantity: str, unit: str, op: str,
+                 name: str, rank_aggr: str = None):
+
+        LogQuantity.__init__(self, name, unit)
+        StateConsumer.__init__(self)
+
+        self.discr = discr
+
+        self.quantity = quantity
+
+        if rank_aggr is None:
+            self.rank_aggr = "norm_2"
+        else:
+            self.rank_aggr = rank_aggr
+
+        from functools import partial
+
+        if op == "min":
+            self._myop = partial(self.discr.nodal_min, "vol")
+        elif op == "max":
+            self._myop = partial(self.discr.nodal_max, "vol")
+        elif op == "sum":
+            self._myop = partial(self.discr.nodal_sum, "vol")
+        else:
+            raise RuntimeError(f"unknown operation {op}")
+
+    @property
+    def default_aggregator(self):
+        """Rank aggregator to use."""
+        if self.rank_aggr == "norm_2":
+            from pytools import norm_2
+            return norm_2
+        elif self.rank_aggr == "norm_inf":
+            from pytools import norm_inf
+            return norm_inf
+        elif self.rank_aggr == "sum":
+            return sum
+        elif self.rank_aggr == "min":
+            return min
+        elif self.rank_aggr == "max":
+            return max
+        else:
+            return None
+
+    def __call__(self):
+        """Return the requested quantity."""
+        raise NotImplementedError
+
+
+class ConservedQuantity(PhysicalQuantity):
+    """Logging support for conserved quantities (mass, energy, momentum)."""
```

Comment:
This code is not in charge of knowing what a conserved quantity is. Refer to the docs for `split_conserved`.